src/sentry/web/frontend/auth_login.py

```diff
@@ -95,39 +243,236 @@ def get_register_form(self, request: Request, initial=None):
+        from sentry.utils.hashlib import md5_text
 
-    def _handle_login(self, request: Request, user, organization: Optional[Organization]):
+        attempted_login = (
+            op == "login" and request.POST.get("username") and request.POST.get("password")
+        )
+
+        return attempted_login and ratelimiter.is_limited(
+            "auth:login:username:{}".format(
+                md5_text(login_form.clean_username(request.POST["username"])).hexdigest()
+            ),
+            limit=5,
+            window=60,  # 5 per minute should be enough for anyone
+        )
+
+    def get_ratelimited_login_form(
+        self, request: Request, login_form: AuthenticationForm, op: str, **kwargs
+    ) -> HttpResponse:
+        """Returns a login form with ratelimited errors displayed."""
+        login_form.errors["__all__"] = [
+            "You have made too many login attempts. Please try again later."
+        ]
+        metrics.incr("login.attempt", instance="rate_limited", skip_internal=True, sample_rate=1.0)
+
+        context = {
+            "op": op or "login",
+            "login_form": login_form,
+        }
+
+        context.update(additional_context.run_callbacks(request))
+        return self.respond_login(request, context, **kwargs)
+
+    def redirect_to_next_login_step(
+        self,
+        request: Request,
+        login_form: AuthenticationForm,
+        organization: Organization,
+    ) -> HttpResponseRedirect:
+        """Called when a user submits a valid login form and must be redirected to
+        the next step in their auth process"""
+        user = login_form.get_user()
+
+        self._handle_login(request, user, organization)
+        metrics.incr("login.attempt", instance="success", skip_internal=True, sample_rate=1.0)
+
+        if not user.is_active:
+            return self.redirect(reverse("sentry-reactivate-account"))
+        if organization:
+            self.refresh_organization_status(request, user, organization)
+        # On login, redirect to onboarding
+        if self.active_organization:
+            if features.has(
+                "organizations:customer-domains",
+                self.active_organization.organization,
+                actor=user,
+            ):
+                setattr(request, "subdomain", self.active_organization.organization.slug)
+        return self.redirect(get_login_redirect(request))
+
+    def _handle_login(
+        self, request: Request, user: User, organization: Optional[Organization]
+    ) -> None:
+        """Logs a user in and determines their active org."""
         login(request, user, organization_id=coerce_id_from(organization))
         self.determine_active_organization(request)
 
-    def handle_basic_auth(self, request: Request, **kwargs):
+    def refresh_organization_status(
+        self, request: Request, user: User, organization: Organization
+    ) -> None:
+        """I'm not really sure what this does actually."""
```

Comment:
So the anwer to that is in the comment on he following line, and above on 384.

In general, organization context is determined at the start of our auth process with a ton of side effects.  In the case of completing login, this organization context is recalculated in a similar, *but not identical* way in order to inform a ton of other interactions that happen.

Yeah, it's a mess.  Tests do generally cover it however.